File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-4.java
class Solution {
    
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        
        for(int i=n-3; i>=0; i--){
            cost[i] += Math.min(cost[i+1], cost[i+2]);
        }
        return Math.min(cost[0], cost[1]);
    }
    // int dfs(int[] cost, int i){
    //     if(i >= cost.length) return 0;

    //     if(dp[i] != -1) return dp[i];

    //     return dp[i] = cost[i] + Math.min(dfs(cost, i+1), dfs(cost, i+2));
    // }
}

File: Data Structures & Algorithms/edit-distance/submission-4.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        dp = [[0] * (n+1) for _ in range(m+1)]
        
        for i in range(m+1):
            dp[i][n] = m-i
        
        for j in range(n+1):
            dp[m][j] = n-j
        
        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                if word1[i] == word2[j]:
                    dp[i][j] = dp[i+1][j+1]
                else:
                    res = min(dp[i+1][j], dp[i][j+1])
                    res = min(res, dp[i+1][j+1])
                    dp[i][j] = res+1
        return dp[0][0]




        dp = {}

        def dfs(i, j):
            if j == n:
                return m-i
            if i == m:
                return n-j
            if (i, j) in dp:
                return dp[(i, j)]
            if word1[i] == word2[j]:
                dp[(i, j)] = dfs(i+1, j+1)
            else:
                res = min(dfs(i, j+1), dfs(i+1, j))
                res = min(res, dfs(i+1, j+1))
                dp[(i, j)] = res+1
            return dp[(i, j)]
        
        return dfs(0,0)
